src/storage_fuzz.zig

```diff
@@ -0,0 +1,164 @@
+const std = @import("std");
+const assert = std.debug.assert;
+const log = std.log.scoped(.fuzz_storage);
+
+const stdx = @import("stdx.zig");
+const vsr = @import("vsr.zig");
+const constants = @import("constants.zig");
+const IO = @import("testing/io.zig").IO;
+const Storage = @import("storage.zig").Storage(IO);
+const fuzz = @import("testing/fuzz.zig");
+
+pub fn main(args: fuzz.FuzzArgs) !void {
+    const zones: []const vsr.Zone = &.{
+        .superblock,
+        .wal_headers,
+        .wal_prepares,
+        .client_replies,
+    };
+
+    const sector_count = 64;
+    const storage_size = sector_count * constants.sector_size;
+    const iterations = args.events_max orelse 10_000;
+
+    const failed_sector_cluster_count = 7;
+    const failed_sector_cluster_minimum_length = 1;
+    const failed_sector_cluster_maximum_length = 4;
+
+    var data_prng = std.rand.DefaultPrng.init(args.seed);
+
+    var fault_map = std.bit_set.ArrayBitSet(u8, sector_count).initEmpty();
+
+    for (0..failed_sector_cluster_count) |_| {
+        const start = data_prng.random().intRangeLessThan(
+            usize,
+            0,
+            sector_count - failed_sector_cluster_maximum_length,
+        );
+        const end = start + data_prng.random().intRangeAtMost(
+            usize,
+            failed_sector_cluster_minimum_length,
+            failed_sector_cluster_maximum_length,
+        );
+
+        fault_map.setRangeValue(.{ .start = start, .end = end }, true);
+    }
+
+    var storage_data_written: [storage_size]u8 align(constants.sector_size) = undefined;
+    @memset(&storage_data_written, 0);
+
+    for (0..sector_count) |sector| {
+        if (!fault_map.isSet(sector)) {
+            data_prng.fill(
+                storage_data_written[sector * constants.sector_size ..][0..constants.sector_size],
+            );
+        }
+    }
+
+    var storage_data_stored: [storage_size]u8 align(constants.sector_size) = undefined;
+    @memset(&storage_data_stored, 0);
+    var storage_data_read: [storage_size]u8 align(constants.sector_size) = undefined;
+    @memset(&storage_data_read, 0);
+
+    var io = IO.init(&.{
+        .{
+            .buffer = &storage_data_stored,
+            .fault_map = &fault_map.masks,
+        },
+    }, .{
+        .seed = args.seed,
+        .larger_than_logical_sector_read_fault_probability = 10,
+    });
+
+    var storage = try Storage.init(&io, 0);
+
+    var write_completion: Storage.Write = undefined;
+
+    for (zones) |zone| {
+        storage.write_sectors(
+            struct {
+                fn callback(completion: *Storage.Write) void {
+                    _ = completion;
+                }
+            }.callback,
+            &write_completion,
+            storage_data_written[zone.start()..][0..zone.size().?],
+            zone,
+            0,
+        );
+
+        storage.tick();
+    }
+
+    for (0..iterations) |_| {
```

Comment:
Let's encompass the writes and generation here too, so the fuzzer tests different fault maps for each iteration.